### ISSUE TITLE
Build bundle for modern browsers and legacy browser separately

### ISSUE THREAD

**Issue by radeno** [zeit/next.js#3056]
- [x] I have searched the [issues](https://github.com/zeit/next.js/issues) of this repository and believe that this is not a duplicate.

There is about 47% desktop browsers and about 45% mobile browsers run ES2015 without any problem. By https://www.netmarketshare.com

There is new simplest solution yet how to deploy separate javascript to modern browsers and transpiled to older browsers. And it is use to <script type="module">
Exists two nice articles how to use it
https://philipwalton.com/articles/deploying-es2015-code-in-production-today/
https://www.contentful.com/blog/2017/04/04/es6-modules-support-lands-in-browsers-is-it-time-to-rethink-bundling/

Both articles explains how to do that with webpack and script tag.
This could be really nice feature, because half of all visitors don't need to transpile to older code.

What we need to make it work?
1. babel-plugin-transform-es2015-modules-commonjs no needed anymore
2. Replace UglifyJs with Babel Minify (Babili)
3. Create new webpack entry and common plugins
4. Set two different babelrc settings for ES2015 and for legacy
5. Create two different script tags explained in blog articles

**Comment by alubbe** [comment]
Step 2 might not be necessary, the newer version of uglify is fully es6 compatible.

Any news on the progress of this issue? We're targeting only recent browsers and would love to lose the 10kb core-js, 2.5kb regenerator and 1kb babel-runtime deps + the size improvements by having es6 over es5 in all files.

**Comment by jasonbarry** [comment]
Updates on this? Some time has passed and there's more browser support across the board. The latest versions of Chrome, Safari, Firefox, Edge, iOS Safari, and Chrome for Android support `type="module"` and the `nomodule` attribute [according to caniuse](https://caniuse.com/#search=nomodule).

Looks like [@babel/preset-env](https://babeljs.io/docs/en/next/babel-preset-env) has an option for module output:
```JS
"presets": [
  ["@babel/preset-env", {
    "targets": {
      "esmodules": true
    }
  }]
]
```

And uglify has a [compress option](https://github.com/mishoo/UglifyJS2/tree/harmony#compress-options) to specify version output:
```JS
new UglifyJsPlugin({
  uglifyOptions: {
    compress: {
      ecma: 6,
    }
  }
});
```

[Quick demo of module/nomodule](https://cdn.rawgit.com/jakearchibald/6110fb6df717ebca44c2e40814cc12af/raw/7fc79ed89199c2512a4579c9a3ba19f72c219bd8/) (view source)

Very much looking forward to the day when Next can let the browser decide which Ecma version to download!

**Comment by timneutkens** [comment]
> Updates on this?
In general, when there's no comment there's no update.

**Comment by JonatanSalas** [comment]
@timneutkens  I imagine that making support for this implies big changes on the Next.JS codebase. Am I wrong? 

@radeno wouldn't be better if instead of using `<script type="module">` to load the generated bundle for modern browsers, and `<script nomodule>` for legacy, just add paths for the next.js server and differ which bundles need to be loaded based on the user-agent? (That's how meteor 1.7 works currently)

**Comment by alexparish** [comment]
@JonatanSalas The user-agent solution would not be viable if your Next.JS server was behind a caching layer like a CDN.

**Comment by JonatanSalas** [comment]
@alexparish yeah, but using the approach explained [here](https://philipwalton.com/articles/deploying-es2015-code-in-production-today/) will lead to issues too except in the case that you can make a custom babel config to support transpiling those unsupported features.

@alexparish would you mind to explain me this? 

> Or at least the cache key would need to contain the user-agent which could cause problems for some use cases.

Why you consider it should cause some problems using the user-agent as the cache-key and in which cases?

**Comment by alexparish** [comment]
@JonatanSalas Sure! I assumed your proposal would involve the Node server rendering the appropriate script tag for the user agent. If you have a caching solution in front of the Node server then the first version of the document will get cached and all user agents will be sent the same HTML and therefore script tag. To work round this any cache providers used would need to be configured appropriately. This is a burden that I don't think Next.js should put on its users.

**Comment by JonatanSalas** [comment]
@alexparish I see your point! Thanks for the explanation.

**Comment by jasonbarry** [comment]
Nuxt.js recently merged this feature in: https://github.com/nuxt/nuxt.js/pull/4231

**Comment by disintegrator** [comment]
@timneutkens are you open to PRs introducing this enhancement?

**Comment by timneutkens** [comment]
Not yet, I have a few ideas on tackling this.

**Comment by fabb** [comment]
There are a few pitfalls:
* module scripts will always be loaded deferred, this might e.g. make polyfills available only after the first executed code that depends on them
* Safari 10.1 supports modules, but will still execute scripts with the nomodule attribute, there is a [fix for this](https://gist.github.com/samthor/64b114e4a4f539915a95b91ffd340acc) though
* using `esmodules: true` in babelrc will disable the browserslist integration - so even when my browserslistrc would exclude some browsers that support modules, that would be overridden, and babel would apply some unnecessary transformations, and if `useBuiltIns: usage` is used, unnecessary polyfills will be pulled in additionally

**Comment by azizhk** [comment]
Hi all,
Can we start by providing a feature to remove polyfills from next.js commons bundle.
Right now what happens is that next is published to npm transpiled. So:
https://github.com/zeit/next.js/blob/d9c462bf379709db8070ff19217b12149f97b63f/packages/next/client/index.js#L18-L20
gets transpiled as 
```js
var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");
var _promise = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/promise"));
if (!window.Promise) {
  window.Promise = _promise.default;
}
```
And this gets uploaded to npm registry. And now there is no way of removing the dependency on `@babel/runtime-corejs2/core-js/promise`. (I tried to pass various`preset-env` options to `next/babel` but then this is hardcoded into the npm registry so can't do anything. 😭 )

**Comment by timneutkens** [comment]
Related RFC: #7563

**Comment by timneutkens** [comment]
Fixed in https://github.com/zeit/next.js/pull/7704

**Comment by balazsorban44** [comment]
This issue has been automatically locked due to no recent activity. If you are running into a similar issue, please create a new issue with the steps to reproduce. Thank you.
